omp/factorization/par_ilu_kernels.cpp

```diff
@@ -50,29 +50,35 @@ namespace par_ilu_factorization {
 
 
 template <typename ValueType, typename IndexType>
-void compute_nnz_l_u(std::shared_ptr<const OmpExecutor> exec,
-                     const matrix::Csr<ValueType, IndexType> *system_matrix,
-                     size_type *l_nnz, size_type *u_nnz)
+void initialize_row_ptrs_l_u(
+    std::shared_ptr<const OmpExecutor> exec,
+    const matrix::Csr<ValueType, IndexType> *system_matrix,
+    IndexType *l_row_ptrs, IndexType *u_row_ptrs)
 {
     auto row_ptrs = system_matrix->get_const_row_ptrs();
     auto col_idxs = system_matrix->get_const_col_idxs();
-    *l_nnz = 0;
-    *u_nnz = 0;
-    for (size_type row = 0; row < system_matrix->get_size()[1]; ++row) {
+    size_type l_nnz{};
+    size_type u_nnz{};
+
+    l_row_ptrs[0] = 0;
+    u_row_ptrs[0] = 0;
+    for (size_type row = 0; row < system_matrix->get_size()[0]; ++row) {
```

Comment:
Maybe we should parallelize this? The nonzeros in the distinct rows are counted in parallel, a second step generates the row pointer out of these counts.

Reply:
We can't simply make the outer loop parallel since each step depends on the other. I will test what benefits we get if we do the whole `for` loop in parallel and do a prefix-sum reduction in the end.

Reply:
Tests are done, the parallel implementation is faster (and now pushed).